### ISSUE TITLE
Suppport "MATERIALIZED" and "NOT MATERIALIZED" modifiers on WITH/CTE

### ISSUE THREAD

**Issue by newhouse** [knex/knex#4470]
# Environment

Knex version: 0.21.12 (yes, old I know)
Database + version: Postgres 12.6
OS: Mac / Linux

# Feature discussion / request

I have a complex query with several CTEs that's something like this:
```js
qb
  .with('foo', function (qb) { ... }
  .with('bar', function (qb) { ... }
  .with('baz', function (qb) { ... }
  .select(...)
  .from('bop')
  .join('baz', 'baz.bop_id', 'bop.id')
```

Having logged the generated SQL, I have debugged a slow query case to being fixed by adding the `MATERIALIZED` keyword to the `baz` CTE.
Changing from `WITH ..., "baz" AS (...) ...` to `WITH ..., "baz" AS MATERIALIZED (...) ...`.

However, it is not clear to me that this is possible with Knex. I've extensively browsed the documentation, and taken a look at the code and it does not look like it's possible to add a keyword like that via the QueryBuilder.

It would be great if this was supported. Perhaps via an optional 3rd parameter to the `with` method like:
```js
// One option: whatever the user types goes in there
qb.with('foo', function (qb) {...}, { modifier: 'MATERIALIZED' })

// Another option: support `materialized` and `notMaterialzied` booleans that add it to the query
qb.with('foo', function (qb) {...}, { materialized: true /* notMaterialized: false  /* })

// And finally:
qb.withMaterialized('foo', function (qb) {...})
qb.withNotMaterialized('foo', function (qb) {...})
```

What do you think?

Also, in the meantime, I could really use a workaround!

Is there any way to write part of a query using raw SQL? It's just 1 of the CTEs that I need to add this keyword to.

Or is there a way to convert the query to raw SQL and perform some kind of replacement to the query before it gets executed? Even though this feels hacky, I'd totally do it .

**Comment by mauriop** [comment]
Really useful if this is going to be supported!
@newhouse what is the workaround that you followed for it?

**Comment by newhouse** [comment]
> Really useful if this is going to be supported!
> @newhouse what is the workaround that you followed for it?

I eventually homed in on this explanation in the docs that pointed out that, starting with Postgres 12, the rules about "materializtion" of CTEs changed: https://www.postgresql.org/docs/12/release-12.html

In particular, I found this line interesting:
 > Specifically, CTEs are automatically inlined if they have no side-effects, are not recursive, and are referenced only once in the query.

Since knex would not currently let me add the `MATERIALIZED` modifier/keyword, I focused on the `"referenced only once"` statement, and triggered the materialization by joining (otherwise pointlessly) to the troublesome CTE a 2nd time. Something like this:

```js
qb
  .with('foo', function (qb) { ... }
  .with('bar', function (qb) { ... }
  .with('baz', function (qb) { ... }
  .select(...)
  .from('bop')
  .join('baz', 'baz.bop_id', 'bop.id')
  // This is the key to the MATERIALIZATION here, even though I don't select anything different
  .join('baz as decoy', 'decoy.bop_id', 'bop.id')
```

For my query, this worked. I hope it does for you, too.

**Comment by twk** [comment]
This would be really useful for us!

**Comment by OlivierCavadenti** [comment]
As I can see, this keyword is supported only by Postgresql and Sqlite ?

**Comment by OlivierCavadenti** [comment]
I am ok to implement this, it's not hard, but generally we like implements things that can be used in more than 2 dialects.

**Comment by kibertoad** [comment]
As long as it is not an intrusive change, I'm ok with supporting only on two dialects, considering that one of them is PostgreSQL, which is a goto RDBMS in Node.js world.

**Comment by newhouse** [comment]
Thanks, everyone!

**Comment by twk** [comment]
Thank you!

**Comment by kibertoad** [comment]
Released in 1.0.2
